File: src/proofs/cxx-inheritence/exe.cc
// -*- mode: Dakota; c-basic-offset: 2; tab-width: 2; indent-tabs-mode: nil -*-

#include <stdio.h>

struct a_t {
  int a;
};
struct b_t : a_t {
  // a
  int b;
};
struct c_t : b_t {
  // a
  // b
  int c;
};
struct d_t {
  int d;
};
struct e_t {
  int e;
};
struct f_t : d_t, e_t {
  // d
  // e
  int f;
};
int main() {
  c_t c {};
  c.a = 0;
  c.b = 0;
  c.c = 0;

  f_t f {};
  f.d = 0;
  f.e = 0;
  f.f = 0;

  printf("sizeof(a_t)=%lu\n", sizeof(a_t));
  printf("sizeof(b_t)=%lu\n", sizeof(b_t));
  printf("sizeof(c_t)=%lu\n", sizeof(c_t));

  printf("sizeof(f_t)=%lu\n", sizeof(f_t));
  
  return 0;
}

File: test/should-pass/pass/aaa/exe.pl
#!/usr/bin/perl -w
# -*- mode: cperl -*-
# -*- cperl-close-paren-offset: -2 -*-
# -*- cperl-continued-statement-offset: 2 -*-
# -*- cperl-indent-level: 2 -*-
# -*- cperl-indent-parens-as-block: t -*-
# -*- cperl-tab-always-indent: t -*-

use strict;
use warnings;

my $aggregate_slots_types = [
  '',
  'struct',
 #'struct : base_t',
 #'struct : base1_t, base2_t',
  'enum',
  'enum : int-t',
  'enum : uint8-t',
  'union',
];
my $sample_primitive_or_typedef_types = [
  'char8-t[64]',
  'object-t (*)(object-t, uint32-t)',
  'int32-t**'
];
my $decl_defn = [
 #'; ',
  '{}'
];
&start(\@ARGV);
sub start {
  my ($argv) = @_;

  print "module dakota;\n\n";
  my $max_len = &max_len($aggregate_slots_types) + 1;
  my $n = 0;
  foreach my $tail (@$decl_defn) {
    foreach my $slots_type (@$aggregate_slots_types) {
      my $pad = &pad($max_len, $slots_type);
      my $in = "slots " . $slots_type . $pad . $tail; 
      my $klass_defn = sprintf("klass klass-%02i { %s }", $n, $in);
      print $klass_defn . "\n";
      $n++;
    }
  }
  print "\n";
  foreach my $slots_type (@$sample_primitive_or_typedef_types) {
    my $in = "slots " . $slots_type . ";";
    my $klass_defn = sprintf("klass klass-%02i { %s }", $n, $in);
    print $klass_defn . "\n";
    $n++;
  }
  print "klass aa::bb::cc { slots {} }" . "\n";
}
sub pad {
  my ($max_len, $str) = @_;
  my $result = '';
  my $str_len = length $str;
  $result = ' ' x ($max_len - $str_len);
  return $result;
}
sub max_len {
  my ($seq) = @_;
  my $max = 0;
  foreach my $str (@$seq) {
    my $len = length $str;
    $max = $len if $len > $max;
  }
  return $max;
}
